File: Data Structures & Algorithms/valid-sudoku/submission-2.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Set<Character>> row = new HashMap<>();
        Map<Integer, Set<Character>> col = new HashMap<>();
        Map<String, Set<Character>> box = new HashMap<>();

        for(int i = 0; i < 9; i++){
            row.put(i, new HashSet<>());
            col.put(i, new HashSet<>());
        }

        for(int i = 0; i < 3; i++){
            for(int j = 0; j < 3; j++){
                box.put("" + i + j, new HashSet<>());
            }            
        }

        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                char curr = board[i][j];
                if(curr != '.'){
                    if(row.get(i).contains(curr)){
                        return false;
                    }

                    row.get(i).add(curr);

                    if(col.get(j).contains(curr)){
                        return false;
                    }

                    col.get(j).add(curr);

                    String key = ""+  (i/3) + (j/3);

                    if(box.get(key).contains(curr)){
                        return false;
                    }

                    box.get(key).add(curr);

                }
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-2.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */
    exist(board, word) {
        const ROW = board.length;
        const COL = board[0].length;

        const dfs = (i, j, n, visited) => {

            if(n === word.length) return true;

            if(i < 0 || i >= ROW || j < 0 || j >= COL || board[i][j] != word[n] || visited.has('' + i + j)){
                return false;
            }

            visited.add('' + i + j);
            const res = dfs(i + 1, j, n + 1, visited) || 
                        dfs(i - 1, j, n + 1, visited) || 
                        dfs(i, j + 1, n + 1, visited) || 
                        dfs(i, j - 1, n + 1, visited);
            visited.delete('' + i + j);
            return res;
        }



        for(let i = 0; i < ROW; i++){

            for(let j = 0; j < COL; j++){
                const set = new Set();
                if(dfs(i, j, 0, set)) return true;
            }
        }

        return false;

    }
}
